server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java

```diff
@@ -283,63 +302,101 @@ LeafBucketCollector getLeafCollector(Comparable<BytesRef> value, LeafReaderConte
             throw new IllegalArgumentException("Expected BytesRef, got " + value.getClass());
         }
         BytesRef term = (BytesRef) value;
-        final SortedSetDocValues dvs = docValuesFunc.apply(context);
         if (leafReaderContextChanged) {
-            remapOrdinals(lookup, dvs);
-            leafReaderOrd = context.ord;
+            // use a separate instance for ordinal and term lookups, that is cached per segment
```

Comment:
This comment isn't right for this method. Here, after your refactoring, we don't have a separate instance as there is no iteration anymore.